### PR TITLE
[CELEBORN-678] ShuffleClientImpl::mapperEnded should not consider attemptId

### PR DESCRIPTION
**Pull request by onebox-li**
### What changes were proposed in this pull request?
ShuffleClientImpl::mapperEnded should not consider attemptId, speculation tasks will update attemptId.

### Why are the changes needed?
ditto

### Does this PR introduce _any_ user-facing change?
No 

### How was this patch tested?
Cluster test

### CODE REVIEW

**Comment by RexXiong** [comment]
LGTM, thanks @onebox-li. For the test failure, we can just remove  STAGE_ENDED/MAP_ENDED related code from FlinkShuffleClientImpl, as that would not happen in Flink.

**Comment by onebox-li** [comment]
thanks @RexXiong for replay, I am not sure about that because I find flink batch support Speculative Execution since 1.16. In my env, I just use rss for spark now, so forget flink-client. May I remove all the mapperEnded judge and these two status code?

**Comment by RexXiong** [comment]
> thanks @RexXiong for reply, I am not sure about that because I find flink batch support Speculative Execution since 1.16. In my env, I just use rss for spark now, so forget flink-client. May I remove all the mapperEnded judge and these two status code?

That's all right. For simplicity and Flink may supprt reduce partition later, FlinkShuffleClientImpl can adapt to the new method(also remove attemptId from mapperEnded)

File: client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java [hunk 1]
```diff
@@ -89,9 +89,11 @@ public class ShuffleClientImpl extends ShuffleClient {
   private final Map<Integer, ConcurrentHashMap<Integer, PartitionLocation>> reducePartitionMap =
       JavaUtils.newConcurrentHashMap();
 
-  protected final ConcurrentHashMap<Integer, Set<String>> mapperEndMap =
+  protected final ConcurrentHashMap<Integer, Set<Integer>> mapperEndMap =
```
Comment (pan3793) [inline client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java:92]: could you please add a comment to mention what the key/value is?

Comment (onebox-li) [reply client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java:93]: ok, done

File: client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java [hunk 2]
```diff
@@ -89,9 +89,12 @@ public class ShuffleClientImpl extends ShuffleClient {
   private final Map<Integer, ConcurrentHashMap<Integer, PartitionLocation>> reducePartitionMap =
       JavaUtils.newConcurrentHashMap();
 
-  protected final ConcurrentHashMap<Integer, Set<String>> mapperEndMap =
+  // key: shuffleId, value: Set(mapId)
+  protected final ConcurrentHashMap<Integer, Set<Integer>> mapperEndMap =
       JavaUtils.newConcurrentHashMap();
 
+  protected final Set<Integer> stageEndShuffleSet = ConcurrentHashMap.newKeySet();
```
Comment (pan3793) [inline client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java:96]: ditto for this one

File: client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java [hunk 3]
```diff
@@ -1525,9 +1524,13 @@ public void setupMetaServiceRef(RpcEndpointRef endpointRef) {
     driverRssMetaService = endpointRef;
   }
 
-  protected boolean mapperEnded(int shuffleId, int mapId, int attemptId) {
-    return mapperEndMap.containsKey(shuffleId)
-        && mapperEndMap.get(shuffleId).contains(Utils.makeMapKey(shuffleId, mapId, attemptId));
+  protected boolean mapperEnded(int shuffleId, int mapId) {
+    return (mapperEndMap.containsKey(shuffleId) && mapperEndMap.get(shuffleId).contains(mapId))
+        || stageEnded(shuffleId);
```
Comment (pan3793) [inline client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java:1529]: how about putting `stageEnded(shuffleId)` at first?

**Review by pan3793 — APPROVED** [review]

File: client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java (hunk 2 above)
Comment (onebox-li) [reply client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java:97]: done

File: client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java (hunk 3 above)
Comment (onebox-li) [reply client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java:1530]: I think mapperEnd may be more likely.

**Review by waitinfuture — APPROVED** [review]
LGTM, thanks! Merging to main/0.3

**Pull request closed without merging**